File: LeetCode/src/dp/CoinChangeII.java
package dp;

public class CoinChangeII {
    public int change(int amount, int[] coins) {

        int[] dp = new int[amount + 1];
        dp[0] = 1;
        for (int i = 0; i < coins.length; i++) {
            for (int j = coins[i]; j <= amount; j++) {
                dp[j] += dp[j - coins[i]];
            }
        }

        return dp[amount];
    }

    public static void main(String[] args) {
        CoinChangeII solution = new CoinChangeII();
        // Test case
        int amount = 5;
        int[] coins = {1, 2, 5};
        System.out.println(solution.change(amount, coins)); // Output: 4
    }
}

File: LeetCode_Python/dp/CoinChangeII.py
class CoinChangeII:
    def change(self, amount: int, coins: list[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1

        for coin in coins:
            for i in range(coin, amount + 1):
                dp[i] += dp[i - coin]

        return dp[amount]


# Time Complexity: O(n * m) where n is the number of coins and m is the amount
# Space Complexity: O(n) where n is the amount


if __name__ == "__main__":
    solution = CoinChangeII()

    print("Excepted Output: 4")
    print("Actual Output: ", solution.change(5, [1, 2, 5]))

    print("Excepted Output: 0")
    print("Actual Output: ", solution.change(3, [2]))

    print("Excepted Output: 1")
    print("Actual Output: ", solution.change(10, [10]))
